File: parse_hci_snoop.py
#!/usr/bin/env python3
"""
Parser pliku btsnooz_hci.log.last — wyciąga pakiety BLE Advertising
z Manufacturer ID 0xC200 (49664) czyli komendy sterujące CaDA.

Format btsnooz:
  0-7:   "btsnooz\x00" (magic)
  8:     version
  9-16:  last timestamp (big-endian int64)
  17+:   zlib compressed data

Po dekompresji — pakiety HCI w ODWROTNEJ kolejności:
  Każdy rekord:
    0-3:  original length (big-endian uint32)
    4-7:  included length (big-endian uint32)  
    8-11: packet flags (big-endian uint32)
    12-15: cumulative drops (big-endian uint32)
    16-23: timestamp (big-endian int64, microseconds)
    24+:  HCI packet data
"""

import struct
import zlib
import sys
from collections import defaultdict

def parse_btsnooz(filepath):
    with open(filepath, 'rb') as f:
        data = f.read()
    
    # Check magic
    magic = data[:8]
    print(f"Magic: {magic}")
    
    if magic == b'btsnooz\x00':
        print("Format: btsnooz (compressed)")
        version = data[8]
        last_ts = struct.unpack('>q', data[9:17])[0]
        print(f"Version: {version}, Last timestamp: {last_ts}")
        
        # Decompress
        try:
            decompressed = zlib.decompress(data[17:])
        except zlib.error:
            # Try with wbits
            try:
                decompressed = zlib.decompress(data[17:], -zlib.MAX_WBITS)
            except:
                decompressed = zlib.decompress(data[17:], zlib.MAX_WBITS | 16)
        
        print(f"Decompressed size: {len(decompressed)} bytes")
        return parse_btsnooz_records(decompressed, last_ts)
    
    elif magic == b'btsnoop\x00':
        print("Format: btsnoop (uncompressed)")
        version = struct.unpack('>I', data[8:12])[0]
        datalink = struct.unpack('>I', data[12:16])[0]
        print(f"Version: {version}, Datalink type: {datalink}")
        return parse_btsnoop_records(data[16:])
    
    else:
        print(f"Unknown format, first 32 bytes: {data[:32].hex()}")
        # Try treating entire file as compressed
        try:
            decompressed = zlib.decompress(data)
            print(f"Successfully decompressed: {len(decompressed)} bytes")
            return parse_btsnooz_records(decompressed, 0)
        except:
            print("Could not decompress. Trying raw scan...")
            return scan_raw_for_c200(data)


def parse_btsnooz_records(data, last_ts):
    """Parse btsnooz decompressed records (stored in reverse order)"""
    packets = []
    offset = 0
    
    while offset + 24 <= len(data):
        try:
            orig_len = struct.unpack('>I', data[offset:offset+4])[0]
            inc_len = struct.unpack('>I', data[offset+4:offset+8])[0]
            flags = struct.unpack('>I', data[offset+8:offset+12])[0]
            drops = struct.unpack('>I', data[offset+12:offset+16])[0]
            ts = struct.unpack('>q', data[offset+16:offset+24])[0]
            offset += 24
            
            if inc_len > 65535 or inc_len == 0:
                # Invalid, try to resync
                offset -= 23
                continue
            
            if offset + inc_len > len(data):
                break
            
            pkt_data = data[offset:offset+inc_len]
            offset += inc_len
            
            packets.append({
                'orig_len': orig_len,
                'inc_len': inc_len,
                'flags': flags,
                'drops': drops,
                'timestamp': ts,
                'data': pkt_data
            })
        except Exception as e:
            print(f"Error at offset {offset}: {e}")
            break
    
    # Reverse because btsnooz stores packets in reverse order
    packets.reverse()
    
    print(f"Parsed {len(packets)} HCI packets")
    return extract_advertising_data(packets)


def parse_btsnoop_records(data):
    """Parse standard btsnoop records"""
    packets = []
    offset = 0
    
    while offset + 24 <= len(data):
        try:
            orig_len = struct.unpack('>I', data[offset:offset+4])[0]
            inc_len = struct.unpack('>I', data[offset+4:offset+8])[0]
            flags = struct.unpack('>I', data[offset+8:offset+12])[0]
            drops = struct.unpack('>I', data[offset+12:offset+16])[0]
            ts = struct.unpack('>q', data[offset+16:offset+24])[0]
            offset += 24
            
            if inc_len > 65535 or inc_len == 0:
                break
            
            if offset + inc_len > len(data):
                break
            
            pkt_data = data[offset:offset+inc_len]
            offset += inc_len
            
            packets.append({
                'orig_len': orig_len,
                'inc_len': inc_len,
                'flags': flags,
                'drops': drops,
                'timestamp': ts,
                'data': pkt_data
            })
        except Exception as e:
            print(f"Error at offset {offset}: {e}")
            break
    
    print(f"Parsed {len(packets)} HCI packets")
    return extract_advertising_data(packets)


def extract_advertising_data(packets):
    """Extract BLE advertising data with Manufacturer ID 0xC200"""
    
    c200_packets = []
    all_adv_packets = []
    
    for i, pkt in enumerate(packets):
        data = pkt['data']
        if len(data) < 4:
            continue
        
        # HCI packet type is first byte
        pkt_type = data[0]
        
        # HCI Command packet (0x01) - LE Set Advertising Data (opcode 0x2008)
        if pkt_type == 0x01 and len(data) >= 4:
            opcode = struct.unpack('<H', data[1:3])[0]
            
            # 0x2008 = LE Set Advertising Data
            # 0x2009 = LE Set Scan Response Data  
            # 0x2036 = LE Set Extended Advertising Data
            if opcode in (0x2008, 0x2009, 0x2036):
                all_adv_packets.append({
                    'index': i,
                    'opcode': hex(opcode),
                    'timestamp': pkt['timestamp'],
                    'raw': data.hex(),
                    'data': data
                })
                
                # Search for Manufacturer Specific Data (AD type 0xFF) with Company ID 0xC200
                mfg_data = find_manufacturer_data(data, 0xC200)
                if mfg_data:
                    c200_packets.append({
                        'index': i,
                        'opcode': hex(opcode),
                        'timestamp': pkt['timestamp'],
                        'mfg_data': mfg_data,
                        'raw': data.hex()
                    })
        
        # Also check for raw advertising data patterns
        # Sometimes the structure differs, so do a byte scan
        if b'\xff\x00\xc2' in data or b'\xff\xc2\x00' in data:
            mfg_data = find_manufacturer_data_raw(data)
            if mfg_data and not any(p['index'] == i for p in c200_packets):
                c200_packets.append({
                    'index': i,
                    'opcode': 'raw_scan',
                    'timestamp': pkt['timestamp'],
                    'mfg_data': mfg_data,
                    'raw': data.hex()
                })
    
    print(f"\n{'='*70}")
    print(f"Found {len(all_adv_packets)} total advertising command packets")
    print(f"Found {len(c200_packets)} packets with Manufacturer ID 0xC200")
    print(f"{'='*70}\n")
    
    if all_adv_packets and not c200_packets:
        print("No 0xC200 packets found. Showing first 20 advertising packets for analysis:\n")
        for p in all_adv_packets[:20]:
            print(f"  Packet #{p['index']} | Opcode: {p['opcode']}")
            print(f"    Raw: {p['raw']}")
            # Try to find any manufacturer data
            find_any_manufacturer_data(p['data'])
            print()
    
    # Analyze unique commands
    if c200_packets:
        analyze_commands(c200_packets)
    
    return c200_packets


def find_manufacturer_data(hci_data, company_id):
    """Find Manufacturer Specific Data in HCI advertising data"""
    # Skip HCI header (type + opcode + param_len)
    # For 0x2008: type(1) + opcode(2) + param_total_len(1) + adv_data_len(1)
    
    data = hci_data
    
    # Search through data for AD structures
    # AD structure: length(1) + type(1) + data(length-1)
    for start in range(3, len(data) - 4):
        ad_len = data[start]
        if ad_len < 3 or start + ad_len >= len(data):
            continue
        
        ad_type = data[start + 1]
        
        # 0xFF = Manufacturer Specific Data
        if ad_type == 0xFF and start + 3 < len(data):
            # Company ID is little-endian
            cid = struct.unpack('<H', data[start+2:start+4])[0]
            if cid == company_id:
                mfg_payload = data[start+4:start+1+ad_len]
                return {
                    'company_id': hex(cid),
                    'payload': mfg_payload.hex(),
                    'payload_bytes': list(mfg_payload),
                    'length': len(mfg_payload),
                    'ad_offset': start
                }
    
    return None


def find_manufacturer_data_raw(data):
    """Brute-force search for 0xC200 manufacturer data pattern"""
    # Look for pattern: 0xFF followed by 0x00 0xC2 (little-endian 0xC200)
    for i in range(len(data) - 4):
        if data[i] == 0xFF and data[i+1] == 0x00 and data[i+2] == 0xC2:
            # Check if previous byte could be a valid AD length
            if i > 0:
                ad_len = data[i-1]
                if 3 <= ad_len <= 25:
                    payload = data[i+3:i-1+ad_len+1]  # rest of AD structure after company ID
                    return {
                        'company_id': '0xc200',
                        'payload': payload.hex(),
                        'payload_bytes': list(payload),
                        'length': len(payload),
                        'ad_offset': i-1
                    }
            # Even without valid length, extract next 16 bytes
            payload = data[i+3:i+3+16]
            if len(payload) >= 4:
                return {
                    'company_id': '0xc200',
                    'payload': payload.hex(),
                    'payload_bytes': list(payload),
                    'length': len(payload),
                    'ad_offset': i
                }
    return None


def find_any_manufacturer_data(data):
    """Find any manufacturer specific data in a packet"""
    for i in range(len(data) - 3):
        if data[i] == 0xFF and i > 0:
            ad_len = data[i-1]
            if 3 <= ad_len <= 25:
                cid = struct.unpack('<H', data[i+1:i+3])[0]
                payload = data[i+3:i-1+ad_len+1]
                print(f"    Found MFG Data: Company ID=0x{cid:04X} ({cid}), payload={payload.hex()}")


def scan_raw_for_c200(data):
    """Last resort: scan raw data for 0xC200 patterns"""
    print("\nScanning raw data for 0xC200 / 0x00C2 patterns...")
    
    findings = []
    # Little-endian: 0x00 0xC2
    for i in range(len(data) - 20):
        if data[i] == 0x00 and data[i+1] == 0xC2:
            context = data[max(0,i-5):i+20]
            findings.append({
                'offset': i,
                'context': context.hex(),
                'raw_16_after': data[i+2:i+18].hex()
            })
    
    # Big-endian: 0xC2 0x00
    for i in range(len(data) - 20):
        if data[i] == 0xC2 and data[i+1] == 0x00:
            context = data[max(0,i-5):i+20]
            findings.append({
                'offset': i,
                'context': context.hex(),
                'raw_16_after': data[i+2:i+18].hex()
            })
    
    print(f"Found {len(findings)} occurrences")
    for f in findings[:50]:
        print(f"  Offset {f['offset']}: context={f['context']} | 16 bytes after={f['raw_16_after']}")
    
    return findings


def analyze_commands(c200_packets):
    """Analyze and group unique command patterns"""
    
    unique_commands = defaultdict(list)
    
    for p in c200_packets:
        payload = p['mfg_data']['payload']
        unique_commands[payload].append(p['index'])
    
    print(f"\n{'='*70}")
    print(f"UNIQUE COMMANDS FOUND: {len(unique_commands)}")
    print(f"{'='*70}\n")
    
    # Sort by first appearance
    sorted_commands = sorted(unique_commands.items(), key=lambda x: x[1][0])
    
    for i, (payload, indices) in enumerate(sorted_commands):
        payload_bytes = bytes.fromhex(payload)
        byte_list = [f"0x{b:02X}" for b in payload_bytes]
        
        print(f"Command #{i+1}: (appeared {len(indices)}x, first at packet #{indices[0]})")
        print(f"  Hex:   {payload}")
        print(f"  Bytes: {' '.join(byte_list)}")
        print(f"  Dec:   {', '.join(str(b) for b in payload_bytes)}")
        if len(payload_bytes) >= 2:
            print(f"  Byte0={payload_bytes[0]:3d} (0x{payload_bytes[0]:02X})  Byte1={payload_bytes[1]:3d} (0x{payload_bytes[1]:02X})", end="")
            if len(payload_bytes) >= 3:
                print(f"  Byte2={payload_bytes[2]:3d} (0x{payload_bytes[2]:02X})", end="")
            if len(payload_bytes) >= 4:
                print(f"  Byte3={payload_bytes[3]:3d} (0x{payload_bytes[3]:02X})", end="")
            print()
        print(f"  Packets: {indices[:10]}{'...' if len(indices)>10 else ''}")
        print()
    
    # Try to guess command functions based on patterns
    print(f"\n{'='*70}")
    print("PATTERN ANALYSIS")
    print(f"{'='*70}\n")
    
    if sorted_commands:
        first_payload = bytes.fromhex(sorted_commands[0][0])
        print(f"Payload length: {len(first_payload)} bytes")
        
        # Check for constant bytes across all commands
        all_payloads = [bytes.fromhex(p) for p, _ in sorted_commands]
        if all_payloads:
            min_len = min(len(p) for p in all_payloads)
            print(f"\nByte-by-byte analysis (first {min_len} bytes):")
            for byte_pos in range(min_len):
                values = set(p[byte_pos] for p in all_payloads)
                if len(values) == 1:
                    print(f"  Byte[{byte_pos}] = CONSTANT 0x{list(values)[0]:02X} ({list(values)[0]})")
                else:
                    vals_str = ', '.join(f'0x{v:02X}' for v in sorted(values))
                    print(f"  Byte[{byte_pos}] = VARIABLE ({len(values)} unique values: {vals_str})")


if __name__ == '__main__':
    filepath = sys.argv[1] if len(sys.argv) > 1 else 'btsnooz_hci.log.last'
    print(f"Parsing: {filepath}")
    print(f"{'='*70}\n")
    
    results = parse_btsnooz(filepath)
    
    if not results:
        print("\n⚠️  No manufacturer data with ID 0xC200 found.")
        print("The file may not contain advertising commands, or the format may differ.")
        print("\nTry alternative: Open in Wireshark with filter:")
        print("  bthci_cmd.opcode == 0x2008")
        print("  or: btcommon.eir_ad.entry.company_id == 0xc200")

File: parse_hci_snoop.js
#!/usr/bin/env node
/**
 * Parser btsnoop HCI log (datalink type 1002 = HCI UART / H4)
 * Wyciąga pakiety BLE Advertising z Manufacturer ID 0xC200
 */

const fs = require('fs');

const filepath = process.argv[2] || 'btsnooz_hci.log.last';
console.log(`Parsing: ${filepath}`);

const rawData = fs.readFileSync(filepath);

// btsnoop header: 16 bytes
// "btsnoop\0" (8) + version(4 BE) + datalink(4 BE)
const magic = rawData.subarray(0, 8).toString('ascii');
console.log(`Magic: "${magic.replace('\0', '\\0')}"`);
console.log(`File size: ${rawData.length} bytes`);

if (magic !== 'btsnoop\x00') {
    console.error('Not a btsnoop file!');
    process.exit(1);
}

const version = rawData.readUInt32BE(8);
const datalink = rawData.readUInt32BE(12);
console.log(`Version: ${version}, Datalink type: ${datalink}`);

// Datalink 1002 = HCI UART (H4) — each packet has H4 type byte
// Datalink 2001 = HCI Unencapsulated
// Record format:
//   Original Length (4 BE)
//   Included Length (4 BE)
//   Packet Flags (4 BE)
//     bit 0: 0=sent, 1=received
//     bit 1: 0=data, 1=command/event
//   Cumulative Drops (4 BE) 
//   Timestamp (8 BE) — microseconds since Jan 1, 0001
//   Packet Data (Included Length bytes)

let packets = [];
let offset = 16; // after header

while (offset + 24 <= rawData.length) {
    const origLen = rawData.readUInt32BE(offset);
    const incLen = rawData.readUInt32BE(offset + 4);
    const flags = rawData.readUInt32BE(offset + 8);
    const drops = rawData.readUInt32BE(offset + 12);
    const timestamp = rawData.readBigUInt64BE(offset + 16);
    offset += 24;

    if (incLen === 0 || incLen > 65535 || offset + incLen > rawData.length) {
        console.log(`Invalid packet at offset ${offset - 24}: incLen=${incLen}`);
        break;
    }

    const pktData = Buffer.from(rawData.subarray(offset, offset + incLen));
    offset += incLen;

    const direction = (flags & 1) === 0 ? 'sent' : 'received';
    const isCommand = (flags & 2) !== 0;

    packets.push({
        origLen, incLen, flags, direction, isCommand,
        timestamp, data: pktData
    });
}

console.log(`\nParsed ${packets.length} HCI packets`);

// Count packet types
const typeCounts = {};
for (const p of packets) {
    const type = p.data.length > 0 ? p.data[0] : -1;
    const key = `0x${type.toString(16).padStart(2, '0')} (${p.direction})`;
    typeCounts[key] = (typeCounts[key] || 0) + 1;
}
console.log('\nPacket type breakdown:');
for (const [key, count] of Object.entries(typeCounts).sort((a, b) => b[1] - a[1])) {
    const typeNames = {
        '0x01': 'HCI Command',
        '0x02': 'ACL Data',
        '0x03': 'SCO Data',
        '0x04': 'HCI Event',
        '0x05': 'ISO Data'
    };
    const hex = key.split(' ')[0];
    const name = typeNames[hex] || 'Unknown';
    console.log(`  ${key} ${name}: ${count}`);
}

// === EXTRACT ADVERTISING COMMANDS ===
console.log('\n' + '='.repeat(70));
console.log('SEARCHING FOR BLE ADVERTISING COMMANDS');
console.log('='.repeat(70) + '\n');

const advPackets = [];
const c200Packets = [];

for (let i = 0; i < packets.length; i++) {
    const p = packets[i];
    const data = p.data;
    if (data.length < 4) continue;

    const h4Type = data[0]; // H4 packet type indicator

    // H4 type 0x01 = HCI Command
    if (h4Type === 0x01 && data.length >= 4) {
        const opcode = data.readUInt16LE(1);
        const paramLen = data[3];

        // LE Set Advertising Data: opcode 0x2008
        // LE Set Scan Response Data: opcode 0x2009
        // LE Set Extended Advertising Data: opcode 0x2037
        if (opcode === 0x2008 || opcode === 0x2009 || opcode === 0x2037) {
            advPackets.push({
                index: i,
                opcode: '0x' + opcode.toString(16).padStart(4, '0'),
                paramLen,
                timestamp: p.timestamp,
                direction: p.direction,
                data
            });

            // Parse advertising data within the command
            // Format: H4(1) + opcode(2) + paramLen(1) + advDataLen(1) + advData(N)
            if (data.length >= 5) {
                const advDataLen = data[4];
                const advData = data.subarray(5, 5 + advDataLen);

                // Parse AD structures within advData
                let adOffset = 0;
                while (adOffset + 2 <= advData.length) {
                    const adLen = advData[adOffset];
                    if (adLen === 0 || adOffset + 1 + adLen > advData.length) break;

                    const adType = advData[adOffset + 1];

                    // 0xFF = Manufacturer Specific Data
                    if (adType === 0xFF && adLen >= 3) {
                        const companyId = advData.readUInt16LE(adOffset + 2);
                        const mfgPayload = advData.subarray(adOffset + 4, adOffset + 1 + adLen);

                        if (companyId === 0xC200) {
                            c200Packets.push({
                                index: i,
                                opcode: '0x' + opcode.toString(16).padStart(4, '0'),
                                timestamp: p.timestamp,
                                direction: p.direction,
                                companyId: '0xC200',
                                payload: mfgPayload.toString('hex'),
                                payloadBytes: [...mfgPayload],
                                rawPacket: data.toString('hex')
                            });
                        } else {
                            // Log other manufacturer IDs (first few)
                        }
                    }

                    adOffset += 1 + adLen;
                }
            }
        }

        // Also check Extended Advertising opcodes
        // LE Set Extended Advertising Parameters: 0x2036
        // LE Set Extended Advertising Enable: 0x2039
        // LE Set Extended Advertising Data: 0x2037
        if (opcode === 0x2036 || opcode === 0x2039) {
            // Just track these
        }
    }
}

console.log(`Found ${advPackets.length} LE Set Advertising Data commands`);
console.log(`Found ${c200Packets.length} packets with Manufacturer ID 0xC200`);

// If no c200 in adv commands, do brute-force on ALL packets
if (c200Packets.length === 0) {
    console.log('\n⚠️  No 0xC200 in advertising commands. Doing full brute-force scan...\n');

    let bruteCount = 0;
    const bruteResults = [];

    for (let i = 0; i < packets.length; i++) {
        const data = packets[i].data;
        for (let j = 0; j < data.length - 5; j++) {
            // Pattern: <adLen> 0xFF <companyID_LE>
            // Company ID 0xC200 in LE = 0x00 0xC2
            if (data[j + 1] === 0xFF && data[j + 2] === 0x00 && data[j + 3] === 0xC2) {
                const adLen = data[j];
                if (adLen >= 3 && adLen <= 30) {
                    bruteCount++;
                    const payload = data.subarray(j + 4, j + 1 + adLen);
                    bruteResults.push({
                        packetIndex: i,
                        h4Type: data[0],
                        direction: packets[i].direction,
                        offset: j,
                        adLen,
                        payload: payload.toString('hex'),
                        payloadBytes: [...payload],
                        context: data.subarray(Math.max(0, j - 2), Math.min(data.length, j + adLen + 3)).toString('hex')
                    });
                }
            }

            // Also try without AD structure — just raw 0x00 0xC2
            if (data[j] === 0x00 && data[j + 1] === 0xC2) {
                // Check if preceded by 0xFF (manufacturer specific data type)
                if (j > 0 && data[j - 1] === 0xFF) {
                    const payload = data.subarray(j + 2, j + 2 + 16);
                    if (!bruteResults.some(r => r.packetIndex === i && r.offset === j - 2)) {
                        bruteCount++;
                        bruteResults.push({
                            packetIndex: i,
                            h4Type: data[0],
                            direction: packets[i].direction,
                            offset: j,
                            adLen: 'N/A (raw)',
                            payload: payload.toString('hex'),
                            payloadBytes: [...payload],
                            context: data.subarray(Math.max(0, j - 3), Math.min(data.length, j + 20)).toString('hex')
                        });
                    }
                }
            }
        }
    }

    console.log(`Brute-force found ${bruteCount} potential 0xC200 matches\n`);

    for (const r of bruteResults.slice(0, 100)) {
        console.log(`  Packet #${r.packetIndex} | H4=0x${r.h4Type.toString(16)} | ${r.direction} | offset=${r.offset}`);
        console.log(`    AD Length: ${r.adLen}`);
        console.log(`    Payload:  ${r.payload}`);
        console.log(`    Bytes:    ${r.payloadBytes.map(b => '0x' + b.toString(16).padStart(2, '0')).join(' ')}`);
        console.log(`    Context:  ${r.context}`);
        console.log();
    }

    // If we found results, analyze them
    if (bruteResults.length > 0) {
        // Convert to c200Packets format
        for (const r of bruteResults) {
            c200Packets.push({
                index: r.packetIndex,
                opcode: 'brute_force',
                timestamp: packets[r.packetIndex].timestamp,
                direction: r.direction,
                companyId: '0xC200',
                payload: r.payload,
                payloadBytes: r.payloadBytes,
                rawPacket: ''
            });
        }
    }
}

// Show first few advertising packets for debugging
if (advPackets.length > 0 && c200Packets.length === 0) {
    console.log('\nFirst 20 advertising data commands (for debugging):\n');
    for (const p of advPackets.slice(0, 20)) {
        console.log(`  #${p.index} | ${p.opcode} | ${p.direction} | paramLen=${p.paramLen}`);
        console.log(`    Full: ${p.data.toString('hex')}`);

        // Parse AD structures
        if (p.data.length >= 5) {
            const advDataLen = p.data[4];
            const advData = p.data.subarray(5, 5 + advDataLen);
            console.log(`    AdvData (${advDataLen} bytes): ${advData.toString('hex')}`);

            let adOff = 0;
            while (adOff + 2 <= advData.length) {
                const adLen = advData[adOff];
                if (adLen === 0) break;
                const adType = advData[adOff + 1];
                const adData = advData.subarray(adOff + 2, adOff + 1 + adLen);
                console.log(`      AD: len=${adLen} type=0x${adType.toString(16).padStart(2, '0')} data=${adData.toString('hex')}`);

                if (adType === 0xFF && adLen >= 3) {
                    const cid = advData.readUInt16LE(adOff + 2);
                    console.log(`        → Manufacturer Data: Company ID = 0x${cid.toString(16).padStart(4, '0')} (${cid})`);
                }

                adOff += 1 + adLen;
            }
        }
        console.log();
    }
}

// === ANALYZE COMMANDS ===
if (c200Packets.length > 0) {
    const uniqueCommands = new Map();
    for (const p of c200Packets) {
        if (!uniqueCommands.has(p.payload)) {
            uniqueCommands.set(p.payload, []);
        }
        uniqueCommands.get(p.payload).push(p.index);
    }

    console.log('\n' + '='.repeat(70));
    console.log(`UNIQUE COMMANDS FOUND: ${uniqueCommands.size}`);
    console.log('='.repeat(70) + '\n');

    const sorted = [...uniqueCommands.entries()].sort((a, b) => a[1][0] - b[1][0]);

    for (let i = 0; i < sorted.length; i++) {
        const [payload, indices] = sorted[i];
        const payloadBuf = Buffer.from(payload, 'hex');
        const byteList = [...payloadBuf].map(b => '0x' + b.toString(16).padStart(2, '0'));

        console.log(`Command #${i + 1}: (appeared ${indices.length}x)`);
        console.log(`  Hex:   ${payload}`);
        console.log(`  Bytes: ${byteList.join(' ')}`);
        console.log(`  Dec:   ${[...payloadBuf].join(', ')}`);
        console.log();
    }

    // Byte-by-byte analysis
    if (sorted.length > 1) {
        const allPayloads = sorted.map(([p]) => Buffer.from(p, 'hex'));
        const minLen = Math.min(...allPayloads.map(p => p.length));

        console.log('='.repeat(70));
        console.log('BYTE-BY-BYTE ANALYSIS');
        console.log('='.repeat(70) + '\n');

        for (let pos = 0; pos < minLen; pos++) {
            const values = new Set(allPayloads.map(p => p[pos]));
            if (values.size === 1) {
                const v = [...values][0];
                console.log(`  Byte[${pos}] CONSTANT = 0x${v.toString(16).padStart(2, '0')} (${v})`);
            } else {
                const vals = [...values].sort((a, b) => a - b);
                console.log(`  Byte[${pos}] VARIABLE (${values.size} unique: ${vals.map(v => '0x' + v.toString(16).padStart(2, '0')).join(', ')})`);
            }
        }
    }

    // JSON output
    console.log('\n' + '='.repeat(70));
    console.log('COMMANDS JSON');
    console.log('='.repeat(70) + '\n');

    const cmdJson = sorted.map(([payload, indices], i) => ({
        id: i + 1,
        hex: payload,
        bytes: [...Buffer.from(payload, 'hex')],
        count: indices.length
    }));
    console.log(JSON.stringify(cmdJson, null, 2));

} else {
    console.log('\n⚠️  No Manufacturer ID 0xC200 data found anywhere in the log.');
    console.log('\nPossible reasons:');
    console.log('  1. The HCI log was captured BEFORE any commands were sent');
    console.log('  2. The advertising data uses a different mechanism not captured in HCI');
    console.log('  3. The file is corrupted or incomplete');
    console.log('\nRecommendation: Use nRF Connect on a second phone to sniff live packets');
    console.log('while pressing buttons in the CaDA app on the first phone.');

    // Let's also dump all unique opcodes we see
    console.log('\n\nAll HCI Command opcodes found:');
    const opcodes = new Map();
    for (const p of packets) {
        if (p.data.length >= 3 && p.data[0] === 0x01) {
            const op = p.data.readUInt16LE(1);
            const key = '0x' + op.toString(16).padStart(4, '0');
            opcodes.set(key, (opcodes.get(key) || 0) + 1);
        }
    }
    const sortedOps = [...opcodes.entries()].sort((a, b) => b[1] - a[1]);
    const bleOpcodeNames = {
        '0x2001': 'LE Set Event Mask',
        '0x2005': 'LE Set Advertising Parameters',
        '0x2006': 'LE Set Advertising Data',
        '0x2007': 'LE Set Scan Response Data',
        '0x2008': 'LE Set Advertising Data',
        '0x2009': 'LE Set Scan Enable',
        '0x200a': 'LE Set Scan Enable',
        '0x200b': 'LE Create Connection',
        '0x200c': 'LE Set Scan Parameters',
        '0x200d': 'LE Set Scan Enable (new)',
        '0x2036': 'LE Set Extended Adv Parameters',
        '0x2037': 'LE Set Extended Adv Data',
        '0x2038': 'LE Set Extended Scan Response Data',
        '0x2039': 'LE Set Extended Adv Enable',
        '0x203e': 'LE Set Extended Scan Parameters',
        '0x203f': 'LE Set Extended Scan Enable',
    };
    for (const [op, count] of sortedOps) {
        const name = bleOpcodeNames[op] || '';
        console.log(`  ${op} ${name}: ${count}`);
    }

    // Dump any LE advertising related commands
    console.log('\n\nDumping LE advertising-related commands:\n');
    const advOps = [0x2005, 0x2006, 0x2007, 0x2008, 0x2009, 0x2036, 0x2037, 0x2038, 0x2039];
    let dumpCount = 0;
    for (let i = 0; i < packets.length && dumpCount < 50; i++) {
        const data = packets[i].data;
        if (data.length >= 3 && data[0] === 0x01) {
            const op = data.readUInt16LE(1);
            if (advOps.includes(op)) {
                dumpCount++;
                console.log(`  Packet #${i} | 0x${op.toString(16).padStart(4, '0')} | ${packets[i].direction}`);
                console.log(`    Full hex: ${data.toString('hex')}`);

                // Try to parse advertising data
                if ((op === 0x2006 || op === 0x2008) && data.length >= 5) {
                    const advLen = data[4];
                    const advData = data.subarray(5, 5 + advLen);
                    console.log(`    Adv data (${advLen}b): ${advData.toString('hex')}`);
                    parseAdStructures(advData);
                }

                if (op === 0x2037 && data.length >= 8) {
                    // Extended: handle(1) + operation(1) + fragment_preference(1) + adv_data_len(1) + data
                    const handle = data[4];
                    const operation = data[5];
                    const fragPref = data[6];
                    const extAdvLen = data[7];
                    const extAdvData = data.subarray(8, 8 + extAdvLen);
                    console.log(`    Ext Adv: handle=${handle} op=${operation} len=${extAdvLen}`);
                    console.log(`    Ext data: ${extAdvData.toString('hex')}`);
                    parseAdStructures(extAdvData);
                }

                console.log();
            }
        }
    }
}

function parseAdStructures(advData) {
    let adOff = 0;
    while (adOff + 2 <= advData.length) {
        const adLen = advData[adOff];
        if (adLen === 0 || adOff + 1 + adLen > advData.length) break;
        const adType = advData[adOff + 1];
        const adData = advData.subarray(adOff + 2, adOff + 1 + adLen);

        const typeNames = {
            0x01: 'Flags',
            0x02: 'Incomplete 16-bit UUID',
            0x03: 'Complete 16-bit UUID',
            0x06: 'Incomplete 128-bit UUID',
            0x07: 'Complete 128-bit UUID',
            0x08: 'Shortened Local Name',
            0x09: 'Complete Local Name',
            0x0A: 'TX Power Level',
            0xFF: 'Manufacturer Specific Data'
        };

        const typeName = typeNames[adType] || 'Unknown';
        console.log(`      AD: len=${adLen} type=0x${adType.toString(16).padStart(2, '0')} (${typeName}) data=${adData.toString('hex')}`);

        if (adType === 0xFF && adLen >= 3) {
            const cid = advData.readUInt16LE(adOff + 2);
            const mfgPayload = advData.subarray(adOff + 4, adOff + 1 + adLen);
            console.log(`        Company ID: 0x${cid.toString(16).padStart(4, '0')} (${cid})`);
            console.log(`        MFG Payload: ${mfgPayload.toString('hex')}`);
            console.log(`        MFG Bytes: ${[...mfgPayload].map(b => b.toString(16).padStart(2, '0')).join(' ')}`);
        }

        adOff += 1 + adLen;
    }
}
